subsys/net/lib/zstream/zstream.c

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2018 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <sys/types.h>
+
+#include <net/zstream.h>
+
+ssize_t zstream_writeall(struct zstream *stream, const void *buf, size_t size,
+			 size_t *written)
+{
+	const char *p = buf;
+	size_t cur_sz = 0;
+
+	do {
+		ssize_t out_sz = zstream_write(stream, p, size);
```

Comment:
shouldn't you do a while() {} instead of a do {} while() here? at least if size is 0 you'll never get into the loop, and it will never allocate out_sz on stack. Here it will do it.

Or then just put out_sz declaration out of the loop, as it will anyway get allocated.

Reply:
> shouldn't you do a while() {} instead of a do {} while() here?

Use of do-while is deliberate, to preserve semantic details of write. The idea is that zstream_writeall() should behave as close as possible to zstream_write() (which is in turn behaves the same as socket write()). So, if user had code like:

~~~
int len = 0;
zstream_write(sock, buf, len);
~~~

Then the underlying socket write (or send()) will be called. The idea is that zstream_writeall() should do the same, for a case when write of length 0 has some special meaning for the underlying object (socket or whatever, as this API isn't limited to sockets).

> and it will never allocate out_sz on stack. Here it will do it.
> Or then just put out_sz declaration out of the loop, as it will anyway get allocated.

We'll be discussing compiler implementation details, but most modern compilers (that's definitely true for gcc and clang) don't allocate anything on entering/leaving C language scopes. Instead, they calculate required stack frame size for the entire function, allocate it once at the beginning of it, and deallocate at the end. That's of course when it will allocate a stack variable at all, because for most vars used in loops, we expect a compiler to be smart to allocate them in registers.

So, we shouldn't worry about machine code quality here - it will be just right, instead should care on about source language, and as "out_sz" is used only inside the body of the loop, it makes sense to define it there, to follow a known best practice of having as small scope for vars as possible.